### PR TITLE
Refactor user sql

### PR DESCRIPTION
**Pull request by seallard**
### Changed
- Refactor user filters.

### This [version](https://semver.org/) is a
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: cg/cli/add.py [hunk 1]
```diff
@@ -93,7 +93,7 @@ def user(context: CGConfig, admin: bool, customer_id: str, email: str, name: str
     """Add a new user with an EMAIL (login) and a NAME (full)."""
     status_db: Store = context.status_db
     customer_obj: models.Customer = status_db.customer(customer_id)
-    existing: models.User = status_db.user(email)
+    existing: models.User = status_db.get_user_by_email(email)
```
Comment (henrikstranneheim) [inline cg/cli/add.py:96]:
```suggestion
    existing_user: User = status_db.get_user_by_email(email=email)
```

File: cg/server/admin.py [hunk 2]
```diff
@@ -21,7 +21,7 @@ class BaseView(ModelView):
     """Base for the specific views."""
 
     def is_accessible(self):
-        user_obj = db.user(session.get("user_email"))
+        user_obj = db.get_user_by_email(session.get("user_email"))
```
Comment (henrikstranneheim) [inline cg/server/admin.py:24]:
```suggestion
        user: User = db.get_user_by_email(email=session.get("user_email"))
```

File: cg/server/admin.py [hunk 3]
```diff
@@ -465,7 +465,7 @@ def cancel_samples(self, entry_ids: List[str]) -> None:
 
     def write_cancel_comment(self, sample: Sample) -> None:
         """Add comment to sample with date and user cancelling the sample."""
-        username: str = db.user(session.get("user_email")).name
+        username: str = db.get_user_by_email(session.get("user_email")).name
```
Comment (henrikstranneheim) [inline cg/server/admin.py:468]:
```suggestion
        user_name: str = db.get_user_by_email(email=session.get("user_email")).name
```

File: cg/server/api.py [hunk 4]
```diff
@@ -80,7 +80,7 @@ def before_request():
             )
         )
 
-    user_obj = db.user(user_data["email"])
+    user_obj = db.get_user_by_email(user_data["email"])
```
Comment (henrikstranneheim) [inline cg/server/api.py:83]:
```suggestion
    user: User = db.get_user_by_email(email=user_data["email"])
```

File: cg/server/invoices/views.py [hunk 5]
```diff
@@ -31,7 +31,7 @@ def before_request():
 
 
 def logged_in():
-    user_obj = db.user(session.get("user_email"))
+    user_obj = db.get_user_by_email(session.get("user_email"))
```
Comment (henrikstranneheim) [inline cg/server/invoices/views.py:34]:
```suggestion
    user: User = db.get_user_by_email(email=session.get("user_email"))
```

File: cg/store/api/findbasicdata.py [hunk 6]
```diff
@@ -144,6 +149,8 @@ def panels(self):
         """Returns all panels."""
         return self.Panel.query.order_by(Panel.abbrev)
 
-    def user(self, email: str) -> User:
-        """Fetch a user from the store."""
-        return self.User.query.filter_by(email=email).first()
+    def get_user_by_email(self, email: str) -> User:
+        """Fetch a user by email from the store."""
```
Comment (henrikstranneheim) [inline cg/store/api/findbasicdata.py:153]:
```suggestion
        """Return a user by email from the database."""
```

File: cg/store/api/findbasicdata.py [hunk 7]
```diff
@@ -144,6 +149,8 @@ def panels(self):
...
+        return apply_user_filter(
+            users=self._get_user_query(), email=email, filters=[UserFilter.FILTER_BY_EMAIL]
```
Comment (henrikstranneheim) [inline cg/store/api/findbasicdata.py:155]:
```suggestion
            users=self._get_user_query(), email=email, filter_filters=[UserFilter.FILTER_BY_EMAIL]
```

File: cg/store/user_filters.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+class UserFilter(Enum):
+    """Define User filter functions."""
+
+    FILTER_BY_EMAIL: Callable = filter_user_by_email
+
+
+def apply_user_filter(
+    users: Query,
+    filters: List[UserFilter],
```
Comment (henrikstranneheim) [inline cg/store/user_filters.py:20]:
```suggestion
    filter_filters: List[Callable],
```
Did we not land on this approach in the stand-up this morning?

File: tests/conftest.py [hunk 9]
```diff
@@ -1736,3 +1736,15 @@ def store_with_multiple_cases_and_samples(
         helpers.add_case_with_sample(base_store=store, case_id=case_id, sample_id=sample_id)
 
     yield store
+
+
+@pytest.fixture(name="store_with_users")
+def fixture_store_with_users(store: Store, helpers: StoreHelpers) -> Store:
+    """Return a store with multiple users."""
+
+    customer = helpers.ensure_customer(store=store)
```
Comment (henrikstranneheim) [inline tests/conftest.py:1745]:
```suggestion
    customer: Customer = helpers.ensure_customer(store=store)
```

File: tests/conftest.py [hunk 10]
```diff
@@ -1736,3 +1736,15 @@ def store_with_multiple_cases_and_samples(
         helpers.add_case_with_sample(base_store=store, case_id=case_id, sample_id=sample_id)
 
     yield store
+
+
+@pytest.fixture(name="store_with_users")
+def fixture_store_with_users(store: Store, helpers: StoreHelpers) -> Store:
```
Comment (henrikstranneheim) [inline tests/conftest.py:1742]: Seems like a nice general feature to add to the base_store.

File: tests/store/api/test_findbasicdata.py [hunk 11]
```diff
@@ -139,3 +149,38 @@ def test_get_collaboration_by_internal_id(base_store: Store, collaboration_id: s
 
     # THEN return a collaboration with the give collaboration internal_id
     assert collaboration.internal_id == collaboration_id
+
+
+def test_get_user_by_email_returns_correct_user(store_with_users: Store):
+    """Test fetching a user by email."""
+
+    # GIVEN a store with multiple users
+    num_users = store_with_users._get_user_query().count()
```
Comment (henrikstranneheim) [inline tests/store/api/test_findbasicdata.py:158]: Add type hints for the tests

File: tests/store/api/test_findbasicdata.py [hunk 12]
```diff
@@ -139,3 +149,38 @@ def test_get_collaboration_by_internal_id(base_store: Store, collaboration_id: s
...
+    assert num_users > 0
+
+    # Select a random user from the store
+    user = store_with_users._get_user_query().first()
+    assert user is not None
+
+    # WHEN fetching the user by email
+    email = user.email
+    filtered_user = store_with_users.get_user_by_email(email=email)
```
Comment (henrikstranneheim) [inline tests/store/api/test_findbasicdata.py:167]:
```suggestion
    filtered_user = store_with_users.get_user_by_email(email=user.email)
```

File: tests/store/api/test_findbasicdata.py [hunk 13]
```diff
@@ -139,3 +149,38 @@ def test_get_collaboration_by_internal_id(base_store: Store, collaboration_id: s
...
+
+    # THEN a user should be returned
+    assert isinstance(filtered_user, User)
+
+    # THEN the email should match
+    assert filtered_user.email == email
```
Comment (henrikstranneheim) [inline tests/store/api/test_findbasicdata.py:173]:
```suggestion
    assert filtered_user.email == user.email
```

File: tests/store/test_user_filters.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def test_filter_user_by_email_returns_correct_user(store_with_users: Store):
+    """Test getting user by email."""
+
+    # GIVEN a store with a user
+    user: User = store_with_users._get_user_query().first()
+    assert user
+
+    user_email: str = user.email
```
Comment (henrikstranneheim) [inline tests/store/test_user_filters.py:13]: Why this extra variable? I think you can use  `user.email` directly

File: tests/store/test_user_filters.py [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+    # THEN the email should match
+    assert filtered_user.email == user_email
+
+
+def test_filter_user_by_email_returns_none_for_nonexisting_email(store_with_users: Store):
+    """Test getting user by email when the email does not exist."""
+
+    # GIVEN a non-existing email
+    non_existent_email = "non_existing@example.com"
```
Comment (henrikstranneheim) [inline tests/store/test_user_filters.py:32]: Used twice in different tests  - make it a fixture

File: tests/store_helpers.py [hunk 16]
```diff
@@ -709,7 +709,7 @@ def ensure_user(
         is_admin: bool = False,
     ) -> models.User:
         """Utility function to add a user that can be used in tests."""
-        user = store.user(email=email)
+        user = store.get_user_by_email(email=email)
```
Comment (henrikstranneheim) [inline tests/store_helpers.py:712]:
```suggestion
        user: User = store.get_user_by_email(email=email)

```

**Review by henrikstranneheim — APPROVED** [review]
Looks great!

**Comment by seallard** [comment]
Deployed to stage and prod

**Pull request merged**